### PR TITLE
Remove unnecessary waiting in ResourceEvent loop [10028]

### PR DESCRIPTION
**Pull request by sfod**
This PR fixes random freezes when large data is transferring using RELIABLE reliability.
Tested for local participants only.

### Details of issue
There is some kind of race condition in ResourceEvent class.
When timer put into pending list, its execution might be delayed depending on what step is being executed in `ResourceEvent::event_service` loop.

For StatefulReader there are two timers that fire periodically and use the same ResourceEvent. One timer is in WriterProxy (`heartbeat_response_`), another is in PDP (`resend_participant_info_event_`).

#### The following sequence shows how issue appears.
Thread 1: calls `heartbeat_response_->restart()`
Thread 1: successfully locks `ResourceEvent::mutex_`.
Thread 2: `ResourceEvent::event_service` waits on `ResourceEvent::mutex_`.
Thread 1: puts `heartbeat_response_` in pending list
Thread 1: releases lock
Thread 2: `ResourceEvent::event_service` calculates the next trigger time as the trigger time of `resend_participant_info_event_` timer.
As a result, data transferring hangs for up to `m_discovery.discovery_config.leaseDuration_announcementperiod` seconds, as there are no answers to heartbeat messages from the affected reader.

This issue appeared on large data (>20 MB) - the larger data the more probability that this issue arises. For example, for 200 MB data chunk it appears in ~90%.
But it seems that it might appear for any kind of data.

The issue this PR fixes looks similar to the following issues:
https://github.com/eProsima/Fast-DDS/issues/1107
https://github.com/eProsima/Fast-DDS/issues/1120
https://github.com/eProsima/Fast-DDS/issues/1062

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3164/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/382/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2824/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3196/)

**Review by MiguelCompany — APPROVED** [review]
Firstly, thank you very much for the analysis you have done!

I have fixed the linter issues and refactored the solution to early continue the loop. I have also added a condition for a possible race when stopping the thread.

It would be great to have a regression test added, but I don't know if it could be done without flakiness. What do you think?

**Review by MiguelCompany — DISMISSED** [review]
I managed to create a test that reproduces the issue.

**Comment by MiguelCompany** [comment]
Waiting for more approvals on this one. @IkerLuengo @richiware

File: test/unittest/rtps/resources/timedevent/TimedEventTests.cpp [hunk 1]
```diff
@@ -333,6 +333,77 @@ TEST(TimedEventMultithread, Event_FourAutoRestart)
...
+
+    // Periodic event that triggers every second
+    auto periodic_callback = []()
+    {
+        return true;
+    };
+    TimedEvent main_event(*env->service_, periodic_callback, 1.0 * periodic_ms.count());
+
+    // Let the periodic event run for several periods
+    main_event.restart_timer();
```
Comment (IkerLuengo) [inline test/unittest/rtps/resources/timedevent/TimedEventTests.cpp:398]:
``` suggestion
    TimedEvent periodic_event(*env->service_, periodic_callback, 1.0 * periodic_ms.count());

    // Let the periodic event run for several periods
    periodic_event.restart_timer();
```

**Review by IkerLuengo — CHANGES_REQUESTED** [review]
@MiguelCompany just a very minor cosmetic change in the name of the event, to be consistent with the rest of the names used in the test.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3178/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/396/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2837/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3210/)

**Comment by sfod** [comment]
Thanks, Miguel, I didn't expect so fast reaction!
I wasn't sure if I should fix linter errors in the affected source file. I'll keep it in mind for the future.
With changes you made the code became much easier to understand, thanks for that.
And as for the test, I couldn't think of a reliable test because of a random nature of the issue. But your test apparently catches it well, very cool.

**Review by IkerLuengo — APPROVED** [review]
LGTM

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3189/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/407/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2848/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3221/)

**Comment by MiguelCompany** [comment]
Current CI looks good to me. I will merge this.

@sfod Thank you very much again!

**Pull request merged**